src/main/java/org/opensearch/flowframework/workflow/AbstractRetryableWorkflowStep.java

```diff
@@ -8,27 +8,133 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Stream;
+
+import static org.opensearch.flowframework.common.CommonValue.REGISTER_MODEL_STATUS;
 import static org.opensearch.flowframework.common.FlowFrameworkSettings.MAX_GET_TASK_REQUEST_RETRY;
 
 /**
  * Abstract retryable workflow step
  */
 public abstract class AbstractRetryableWorkflowStep implements WorkflowStep {
-
+    private static final Logger logger = LogManager.getLogger(AbstractRetryableWorkflowStep.class);
     /** The maximum number of transport request retries */
     protected volatile Integer maxRetry;
+    private final MachineLearningNodeClient mlClient;
+    private final FlowFrameworkIndicesHandler flowFrameworkIndicesHandler;
 
     /**
      * Instantiates a new Retryable workflow step
      * @param settings Environment settings
      * @param clusterService the cluster service
+     * @param mlClient machine learning client
+     * @param flowFrameworkIndicesHandler FlowFrameworkIndicesHandler class to update system indices
      */
-    public AbstractRetryableWorkflowStep(Settings settings, ClusterService clusterService) {
+    public AbstractRetryableWorkflowStep(
+        Settings settings,
+        ClusterService clusterService,
+        MachineLearningNodeClient mlClient,
+        FlowFrameworkIndicesHandler flowFrameworkIndicesHandler
+    ) {
         this.maxRetry = MAX_GET_TASK_REQUEST_RETRY.get(settings);
         clusterService.getClusterSettings().addSettingsUpdateConsumer(MAX_GET_TASK_REQUEST_RETRY, it -> maxRetry = it);
+        this.mlClient = mlClient;
+        this.flowFrameworkIndicesHandler = flowFrameworkIndicesHandler;
+    }
+
+    /**
+     * Retryable get ml task
+     * @param workflowId the workflow id
+     * @param nodeId the workflow node id
+     * @param future the workflow step future
+     * @param taskId the ml task id
+     * @param retries the current number of request retries
+     */
+    void retryableGetMlTask(
+        String workflowId,
+        String nodeId,
+        CompletableFuture<WorkflowData> future,
+        String taskId,
+        int retries,
+        String message
+    ) {
+        mlClient.getTask(taskId, ActionListener.wrap(response -> {
+            MLTaskState currentState = response.getState();
+            if (currentState != MLTaskState.COMPLETED) {
+                if (Stream.of(MLTaskState.FAILED, MLTaskState.COMPLETED_WITH_ERROR).anyMatch(x -> x == currentState)) {
+                    // Model registration failed or completed with errors
+                    String errorMessage = message + " failed with error : " + response.getError();
+                    logger.error(errorMessage);
+                    future.completeExceptionally(new FlowFrameworkException(errorMessage, RestStatus.BAD_REQUEST));
+                } else {
+                    // Task still in progress, attempt retry
+                    throw new IllegalStateException(message + " is not yet completed");
```

Comment:
Avoid using exceptions for normal process flow.  They're a serious performance hit. https://dzone.com/articles/exceptions-as-controlflow-in-java

Why not just do the retry recursion here instead of in a catch block?

I recognize that regular recursion impacts the stack but if you add a `() ->` in the return you can basically achieve tail recursion and avoid it.  See https://github.com/jonckvanderkogel/java-tail-recursion/blob/master/src/main/java/tailrecursion/Factorial.java

But better than that, why not just do this all in a loop?   while... not COMPLETE and tries < retries...

Reply:
@joshpalis has this method. I pulled it in the AbstractClass. @joshpalis WDYT?

Reply:
No strong opinion, we can modify this retry function to just do everything in a loop